### ISSUE TITLE
Using this as a lib or not launching via python script

### ISSUE THREAD

**Issue by Folcon** [HumbleUI/HumbleUI#12]
This may be too early in the project for this to be viable, but I'm trying to see how to start either using this as a lib or using the code in this lib without having to call the python scripts.

I'm using leiningen which might not be the best way to do this.

[Here's the code](https://github.com/Folcon/fruit-economy) which I'm going to try and see if I can get working =)...

```shell
%  lein trampoline repl
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
Compiling 9 source files to ~/Code/clojure/fruit-economy/target/classes
~/Code/clojure/fruit-economy/src/java/module-info.java:4: error: module not found: lombok
    requires static lombok;
                    ^
~/Code/clojure/fruit-economy/src/java/module-info.java:5: error: module not found: org.jetbrains.annotations
    requires static org.jetbrains.annotations;
                                 ^
~/Code/clojure/fruit-economy/src/java/module-info.java:6: error: module not found: clojure
    requires clojure;
             ^
3 errors
Compilation of Java sources(lein javac) failed.
```

Not quite sure what part of the python script I'm missing, hopefully will have a chance this weekend to dig in and find out. Pretty excited about this.

**Comment by tonsky** [comment]
Sorry about that. Eventually I tend to forgot how painful it is to host java sources along with clojure sources and attempt it again.

I did a little bit of reorganization. All java parts have moved to `io.github.humbleui:types:0.1.2:clojure`. This repo is now Clojure-only. This should make including it as a dependency much easier.

For tools.deps this seem to work: `io.github.humbleui/humbleui {:git/sha "7db69df04963a6112837c2f5f8731e58f00bfac3"}`

For Lein, you might now build a jar by using `./script/package.py` and depend on that. Eventually a jar built exactly that way will end up on Clojars, but for now I think it’s too early for that.

Does that work for you? Let me know! I can adjust more things if needed.

**Comment by Folcon** [comment]
Thanks! [I've updated the linked project](https://github.com/Folcon/fruit-economy) =)...

I've used lein-git-down to get the specific git hash, so that's fine, it still seems I need to still import `jwm` `skija-shared`, plus all the `skija` os specific libs, which is no hardship, but useful to know.

Also not sure if this makes a difference, but I imported `io.github.humbleui/types` instead of `io.github.humbleui/types$clojure` as the latter was not available.

The weird thing is that unlike when I run your python script version, running it via leinigen just seems to make it not work, not sure why yet as you can see `-XstartOnFirstThread` is there so it really shouldn't be terminating due to `NSWindow drag regions should only be invalidated on the Main Thread!`... Not sure if you have any insight into what's going on there...

```
Starting nREPL server...
/Library/Java/JavaVirtualMachines/openjdk.jdk/Contents/Home/bin/java -Dfile.encoding=UTF-8 -XstartOnFirstThread -XstartOnFirstThread -Dclojure.compile.path=~/Code/clojure/fruit-economy/target/classes -Dfruit-economy.version=0.1.0-SNAPSHOT -Dclojure.debug=false "-javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=57360:/Applications/IntelliJ IDEA.app/Contents/bin" -classpath ~/Code/clojure/fruit-economy/test:~/Code/clojure/fruit-economy/dev:~/Code/clojure/fruit-economy/src:~/Code/clojure/fruit-economy/resources:~/Code/clojure/fruit-economy/target/classes:~/.m2/repository/io/github/humbleui/skija-macos-arm64/0.98.0/skija-macos-arm64-0.98.0.jar:~/.m2/repository/nrepl/nrepl/0.9.0/nrepl-0.9.0.jar:~/.m2/repository/io/github/humbleui/skija-macos-x64/0.98.0/skija-macos-x64-0.98.0.jar:~/.m2/repository/io/github/humbleui/skija-windows/0.98.0/skija-windows-0.98.0.jar:~/.m2/repository/io/github/humbleui/skija-linux/0.98.0/skija-linux-0.98.0.jar:~/.m2/repository/org/clojure/clojure/1.11.0-alpha3/clojure-1.11.0-alpha3.jar:~/.m2/repository/org/clojure/core.specs.alpha/0.2.62/core.specs.alpha-0.2.62.jar:~/.m2/repository/org/clojure/spec.alpha/0.2.194/spec.alpha-0.2.194.jar:~/.m2/repository/io/github/humbleui/types/0.1.2/types-0.1.2.jar:~/.m2/repository/humbleui/humbleui/7db69df04963a6112837c2f5f8731e58f00bfac3/humbleui-7db69df04963a6112837c2f5f8731e58f00bfac3.jar:~/.m2/repository/io/github/humbleui/jwm/0.3.1/jwm-0.3.1.jar:~/.m2/repository/io/github/humbleui/skija-shared/0.98.0/skija-shared-0.98.0.jar clojure.main -i /private/var/folders/77/2pz9tf451x962w3371g_xhg40000gn/T/form-init10275749412739895747.clj
Reflection warning, io/github/humbleui/ui.clj:259:5 - call to method drawRect on java.lang.Object can't be resolved (no such method).
Error loading complete.core: Could not locate complete/core__init.class, complete/core.clj or complete/core.cljc on classpath.
Connecting to local nREPL server...
Clojure 1.11.0-alpha3
nREPL server started on port 57362 on host 127.0.0.1 - nrepl://127.0.0.1:57362
1
=> 1
(-main)
nREPL server started on port 57373 on host localhost - nrepl://localhost:57373
2022-01-14 12:37:01.165 java[70185:3940439] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff37fe84d7 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff70a2b5bf objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff38010cbc -[NSException raise] + 9
	3   AppKit                              0x00007fff3520dc1c -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 310
	4   AppKit                              0x00007fff351f5682 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1416
	5   AppKit                              0x00007fff351f50f3 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 42
	6   libjwm_x64.dylib                    0x000000010a3b61b7 _ZN3jwm9WindowMac4initEv + 263
	7   libjwm_x64.dylib                    0x000000010a3b663a Java_io_github_humbleui_jwm_WindowMac__1nMake + 138
	8   ???                                 0x0000000112f6a53a 0x0 + 4613121338
)
libc++abi.dylib: terminating with uncaught exception of type NSException

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

Also, the java version I'm running in case that's important:
```
/Library/Java/JavaVirtualMachines/openjdk.jdk/Contents/Home/bin/java -version
openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment Homebrew (build 17.0.1+1)
OpenJDK 64-Bit Server VM Homebrew (build 17.0.1+1, mixed mode, sharing)
```

**Comment by Folcon** [comment]
My mistake, worked out how to handle that `$` difference in deps, so learned something new today =)...

**Comment by tonsky** [comment]
> it still seems I need to still import jwm skija-shared, plus all the skija os specific libs, which is no hardship, but useful to know.

Interesting! These dependencies are supposed to be transitive (see `deps.edn`). lein-git-down claims to support this format, did it not work?

> Also not sure if this makes a difference, but I imported io.github.humbleui/types instead of io.github.humbleui/types$clojure as the latter was not available.

It should, actually. `io.github.humbleui/types$clojure` is tool.deps way of saying `0.1.2-clojure` (classifer, in Maven terms)

> running it via leinigen just seems to make it not work

So here’s the deal. macOS is very strict about only doing UI stuff from first thread. That means you can’t run nREPL and start app from there. What you can is call `-main` that spawns nREPL in separate thread and then call `(hui/init) (hui/start)` on the original one and block forever. See how https://github.com/HumbleUI/HumbleUI/blob/7db69df04963a6112837c2f5f8731e58f00bfac3/dev/user.clj#L132-L136 is structured

Another gotcha with lein is that it starts your app from its own JVM (or something like that), so make sure to run `lein trampoline run` instead of just `lein run`.

Also, don’t forget `-XstartOnFirstThread`, otherwise JVM itself won’t behave.

I know it’s a pain in the ass, but I don’t think there’s a good way around these limitations. It’s much more forgiving on Windows/Linux, but on macOS it is what it is.

Let me know if this helps.

**Comment by Folcon** [comment]
Hmm, so I have a partial solution, though this might just be a limitation with working with a REPL like cursive does.

Specifically I can launch it like you specified in a terminal via `lein with-profile +macos,+dev trampoline run`, but trying to create a REPL that does this still blows up[0].

So it's a little inelegant, but the good thing is that it works, so for the moment I'm happy with it =)... Hacking on things can commence!

I've been able to connect a remote repl to the nrepl instance and work with it as usual, calls such as:

```clojure
(hui/doui (window/set-z-order @*window :normal))
(hui/doui (window/set-z-order @*window :floating))
```

Both appear to work, which is great!

One thing that's interesting is that this however does blow up:
```clojure
(do
    (hui/doui (some-> @*window window/close))
    (reset! *window (hui/doui (make-window))))
```
I'm assuming that it works in if you launch via the python script?

I instead get this:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010b7b2eae, pid=71518, tid=775
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.1+12 (17.0.1+12) (build 17.0.1+12)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (17.0.1+12, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [libjwm_x64.dylib+0xdeae]  Java_io_github_humbleui_jwm_WindowMac__1nRequestFrame+0xe
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# ~/Code/clojure/fruit-economy/hs_err_pid71518.log
Compiled method (n/a)    8502 3756     n 0       io.github.humbleui.jwm.WindowMac::_nRequestFrame (native)
 total in heap  [0x000000011e35aa90,0x000000011e35ade0] = 848
 relocation     [0x000000011e35abf0,0x000000011e35ac20] = 48
 main code      [0x000000011e35ac20,0x000000011e35add8] = 440
 oops           [0x000000011e35add8,0x000000011e35ade0] = 8
Compiled method (c1)    8503 3564       3       io.github.humbleui.ui.Contextual::_draw (66 bytes)
 total in heap  [0x000000011747d810,0x000000011747e3b8] = 2984
 relocation     [0x000000011747d970,0x000000011747da10] = 160
 main code      [0x000000011747da20,0x000000011747e1c0] = 1952
 stub code      [0x000000011747e1c0,0x000000011747e210] = 80
 oops           [0x000000011747e210,0x000000011747e228] = 24
 metadata       [0x000000011747e228,0x000000011747e240] = 24
 scopes data    [0x000000011747e240,0x000000011747e2b0] = 112
 scopes pcs     [0x000000011747e2b0,0x000000011747e380] = 208
 dependencies   [0x000000011747e380,0x000000011747e388] = 8
 nul chk table  [0x000000011747e388,0x000000011747e3b8] = 48
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```


-[0]: Here's a stacktrace of the behaviour if you're interested.

I've run `source` on `-main` so you can see it's the same as yours.

`-XstartOnFirstThread` is on that path and `-main` appears to be the same as well, but it's still blowing up. There's probably some odd threading behaviour that's going on that I've not thought about.

```
Starting nREPL server...
/Library/Java/JavaVirtualMachines/openjdk.jdk/Contents/Home/bin/java -Dfile.encoding=UTF-8 -XstartOnFirstThread -Dclojure.compile.path=~/Code/clojure/fruit-economy/target/classes -Dfruit-economy.version=0.1.0-SNAPSHOT -Dclojure.debug=false "-javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=57690:/Applications/IntelliJ IDEA.app/Contents/bin" -classpath ~/Code/clojure/fruit-economy/test:~/Code/clojure/fruit-economy/dev:~/Code/clojure/fruit-economy/src:~/Code/clojure/fruit-economy/resources:~/Code/clojure/fruit-economy/target/classes:~/.m2/repository/io/github/humbleui/skija-macos-arm64/0.98.0/skija-macos-arm64-0.98.0.jar:~/.m2/repository/nrepl/nrepl/0.9.0/nrepl-0.9.0.jar:~/.m2/repository/io/github/humbleui/skija-macos-x64/0.98.0/skija-macos-x64-0.98.0.jar:~/.m2/repository/io/github/humbleui/types/0.1.2/types-0.1.2-clojure.jar:~/.m2/repository/io/github/humbleui/skija-windows/0.98.0/skija-windows-0.98.0.jar:~/.m2/repository/io/github/humbleui/skija-linux/0.98.0/skija-linux-0.98.0.jar:~/.m2/repository/org/clojure/clojure/1.11.0-alpha3/clojure-1.11.0-alpha3.jar:~/.m2/repository/org/clojure/core.specs.alpha/0.2.62/core.specs.alpha-0.2.62.jar:~/.m2/repository/org/clojure/spec.alpha/0.2.194/spec.alpha-0.2.194.jar:~/.m2/repository/humbleui/humbleui/7db69df04963a6112837c2f5f8731e58f00bfac3/humbleui-7db69df04963a6112837c2f5f8731e58f00bfac3.jar:~/.m2/repository/io/github/humbleui/jwm/0.3.1/jwm-0.3.1.jar:~/.m2/repository/io/github/humbleui/skija-shared/0.98.0/skija-shared-0.98.0.jar clojure.main -i /private/var/folders/77/2pz9tf451x962w3371g_xhg40000gn/T/form-init1262808874139951653.clj
Reflection warning, io/github/humbleui/ui.clj:259:5 - call to method drawRect on java.lang.Object can't be resolved (no such method).
Error loading complete.core: Could not locate complete/core__init.class, complete/core.clj or complete/core.cljc on classpath.
Connecting to local nREPL server...
Clojure 1.11.0-alpha3
nREPL server started on port 57692 on host 127.0.0.1 - nrepl://127.0.0.1:57692
(source -main)
(defn -main [& args]
  (future (apply nrepl/-main args))
  (hui/init)
  (reset! *window (make-window))
  (hui/start))
=> nil
(-main)
nREPL server started on port 57695 on host localhost - nrepl://localhost:57695
2022-01-14 13:13:35.120 java[71069:3968711] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff37fe84d7 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff70a2b5bf objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff38010cbc -[NSException raise] + 9
	3   AppKit                              0x00007fff3520dc1c -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 310
	4   AppKit                              0x00007fff351f5682 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1416
	5   AppKit                              0x00007fff351f50f3 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 42
	6   libjwm_x64.dylib                    0x000000010ca6f1b7 _ZN3jwm9WindowMac4initEv + 263
	7   libjwm_x64.dylib                    0x000000010ca6f63a Java_io_github_humbleui_jwm_WindowMac__1nMake + 138
	8   ???                                 0x000000011544553a 0x0 + 4651767098
)
libc++abi.dylib: terminating with uncaught exception of type NSException

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

**Comment by tonsky** [comment]
It blows up for me too. I’ll investigate

**Comment by tonsky** [comment]
So the quick fix is to replace

```
(some-> @*window (window/request-frame))
```

with

```
(window/request-frame window)
```

As it turned out, `:on-close` is not called when closing window programmatically. Will fix it in the next JWM version, before that, use the fix

**Comment by Folcon** [comment]
Perfect, works as suggested, thanks!

A question, what sorts of things are in / out of scope for this?

I'm happy to create issues as I run into things, but I don't want to just produce noise.

My current goal is to try creating a game interface, basically mix some canvas with UI controls, so I can get something akin to a display area, buttons / controls on the left or right, with some space for status info at the top / bottom of the screen.

So something similar to an intellij window, except instead of the main UI being a text editor, instead rendering some 2d stuff...

**Comment by tonsky** [comment]
Should be okay. It’s very early though, we barely have any widgets yet

**Comment by Folcon** [comment]
That's ok, glad to hear that what I'm trying to do is not out of scope, I'll try and see how far I get by adding them myself...

I'm planning on using the repo linked to this project as a bit of a test-bed to see what's workable...
